documentation/dataplane/dash-flow/DASH_FLOW_SAI.md

```diff
@@ -0,0 +1,615 @@
+rewrite_info.iflow_sport = 12345;
+rewrite_info.iflow_vni = 1002;
+
+/* Assign the rewrite info to the flow metadata */
+flow_metadata.rewrite_info = &rewrite_info;
+
+/* Serialize the protobuf message to bytes */
+unsigned len = sai_dash_flow_metadata__get_packed_size(&flow_metadata);
+uint8_t *buf = malloc(len);
+sai_dash_flow_metadata__pack(&flow_metadata, buf);
+
+sai_attribute_t sai_attrs_list[1];
+sai_attrs_list[0].id = SAI_FLOW_STATE_ATTR_METADATA_PROTOBUF;
+sai_attr_list[0].value = buf;
+
+
+/* Use the buffer as desired, e.g., as an attribute */
+/* ... */
+
+/* Free the allocated buffer */
+free(buf);
+
+```
+
+
+
+## Flow Table Operation
+
+#### Create Flow Table
+
+```c
+/* Attributes for the flow table */
+uint32_t attr_count = ...; // Specify the number of attributes
+sai_attribute_t *attr_list = ...; // Provide the specific array of attributes
+
+/* Create a new flow table */
+sai_status_t status = sai_dash_flow_create_table(&flow_table_id, attr_count, attr_list);
+if (status != SAI_STATUS_SUCCESS) {
+    /* handle the error */
+}
+```
+
+#### Add Flow Entries
+
+```c
+/* Add multiple flow entries to the table in bulk */
+uint32_t flow_count = num_flow_states;
+const sai_dash_flow_key_t flow_key[] = ...; // Provide the specific array of flow keys
+uint32_t attr_count[] = ...; // Specify the number of attributes for each flow
+sai_attribute_t *attr_list[] = ...; // Provide the specific arrays of attributes for each flow
+sai_status_t object_statuses[] = ...; // Buffer to store statuses for each flow
+
+status = sai_dash_flow_bulk_create_entry(flow_table_id, flow_count, flow_key, attr_count, attr_list, SAI_BULK_OP_ERROR_MODE_IGNORE_ERROR, object_statuses);
+if (status != SAI_STATUS_SUCCESS) {
+    /* handle error */
+}
+
+```
+
+#### Retrieve Flow Entries
+
+```c
+/* Retrieve multiple flow entries and call the callback function to further process */
+/* Define the callback function */
+void process_flow_state(sai_object_id_t flow_table_id, const sai_dash_flow_key_t* key[], uint32_t* attr_count[], sai_attribute_t* attr_list[]) {
+    for (uint32_t i = 0; i < attr_count[i]; i++) {
+        /* Process each flow entry in the list... */
+    }
+}
+
+/* Retrieve the flow entries from the table using the specified keys */
+status = sai_dash_flow_bulk_get_entry_callback(flow_table_id, &key, SAI_FLOW_TABLE_ENTRY_QUERY_ALL, &attr_count, &attr_list, process_flow_state, 3 * aging_time, &finish);
+if (status != SAI_STATUS_SUCCESS) {
+    /* handle error */
+}
+/* The callback function process_flow_state will be invoked for each key in flow_keys */
+
+while (!finish) {
+  /* The bulk get is ongoing */
+  sleep(1);
+}
+/* The bulk get is done */
+
+```
+
+#### Remove flow entry
+
+```c
+/* Remove the flow entry from the table */
+status = sai_dash_flow_remove_entry(flow_table_id, &example_flow_state.key);
+if (status != SAI_STATUS_SUCCESS) {
+    /* handle error */
+}
+
+```
+
+#### Remove flow table
+
+```c
+/* Finally, remove the flow table */
```

Comment:
Do you expect this API to actively remove all the entries from the HW before returning the status? The alternative is to do perform a lazy delete where the ENI will be marked with a drop status and we can let HW age-out the entries over time.

Reply:
@zhixiongniu , this is a very point. I believe an attribute to control the flow expiry will be awesome to add.